[weathercompany] Weather Company Binding initial contribution

mhilbush opened the pull request:
The Weather Underground API was turned off in March. However, people who upload their Personal Weather Station (PWS) weather data to Weather Underground have been given access to the weather forecast API from The Weather Company (TWC). TWC API provides access to PWS data, as well as a 5-day weather forecast.

The binding currently supports:
- 5 day forecast (one daily forecast for Today, Tomorrow, Day 2, Day 3, Day 4, Day 5)
- 5 day “daypart” (Day and Night) forecast for each of the above days
- PWS current observations (must set station ID in the thing config)
- TWC’s full list of language translations, which is quite extensive
- UoM for conversion of units
- configurable refresh time of forecast (fastest refresh is 2 minutes, default is 30 minutes)
- location types of postal code (only for US, UK, DE, FR, IT), geocode (latitude and longitude), and IATA airport code
- auto discovery of local weather thing if location and locale are set in openHAB

Community thread where this binding has been discussed is [here](https://community.openhab.org/t/weatherunderground-replacement-for-pws-data-uploaders-only/69122?u=mhilbush).

Signed-off-by: Mark Hilbush mark@hilbush.com

This is ready for review.

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java:
```diff
@@ -0,0 +1,744 @@
+    // Personal Weather Station observations URL
+    private static final String BASE_PWS_URL = "https://api.weather.com/v2/pws/observations/current";
+
+    public static final int WEATHER_COMPANY_API_TIMEOUT = 10000;
```
Is this miliseconds and the next line seconds? Maybe you could use something which makes it more clear for example by giving it a more exact name?

mhilbush replied:
Yes. Both now are represented as seconds.

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java:
```diff
@@ -0,0 +1,744 @@
+    private final Gson gson = new GsonBuilder().serializeNulls().create();
+
+    // Provided by handler factory
+    private @NonNullByDefault({}) TimeZoneProvider timeZoneProvider;
```
Given they compare in through the constructor I would make them final

And please don't use this annotation hack except for References

mhilbush replied:
> Given they compare in through the constructor I would make them final

Good catch.

> And please don't use this annotation hack except for References

Ok. I had to rework a bit of the code to resolve another usage of this.

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java:
```diff
@@ -0,0 +1,744 @@
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
```
AFAIK this is the default already?

mhilbush replied:
Right. This should've been OFFLINE.

martinvw replied:
Why would you mark it OFFLINE at first, its not correct IMHO? What about just leaving it UNKNOWN until you know the status?

mhilbush replied:
Because it's not ONLINE yet? 😉 

BTW, the [archetype](https://github.com/openhab/openhab-core/blob/master/tools/archetype/binding/src/main/resources/archetype-resources/src/main/java/internal/__bindingIdCamelCase__Handler.java#L80) explicitly sets the status to UNKNOWN as one of the first things it does in initialization.

I can remove it completely, or leave the call to set it to UNKNOWN (which is consistent with the archetype).

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java:
```diff
@@ -0,0 +1,744 @@
+
+        // Get the configuration
+        config = getConfigAs(WeatherCompanyConfig.class);
+        // logger.debug("Handler: Config: apiKey={}", config.apiKey);
```
Would it make sense to log the String representation of the configuration and solve the rest there?

mhilbush replied:
Yes, but... I didn't want to log the api key. I've removed the logging of the config completely.

martinvw replied:
Indeed, but my intention was to embed that logic as well in the toString implementation of the configuration :-)

mhilbush replied:
Yep, that'll work.

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java:
```diff
@@ -0,0 +1,744 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // Handler doesn't handle any commands
```
You should handle the refresh command, optionally with some cache so that you don't overload the backend.

martinvw replied:
Is Github right that this is still open?

mhilbush replied:
I have a solution for it. I just didn't put an update here yet.

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java:
```diff
@@ -0,0 +1,744 @@
+                } else {
+                    sb.append("&iataCode=").append(config.iataCode.replace(" ", "").toUpperCase());
+                }
+                break;
```
`default`? not valid?

mhilbush replied:
Default is not valid. I added the default case.

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java:
```diff
@@ -0,0 +1,744 @@
+        // Set API key from config
+        sb.append("&apiKey=").append(config.apiKey);
+        // Set the units to Imperial or Metric
+        sb.append("&units=").append(getUnitsQueryString());
```
Why do you pass the unit here but not on line 245 is that something which should be done, or are they both correct?

mhilbush replied:
Good spot! This was intentional.

For the PWS observations, it's sufficient to request Imperial units, then let UoM convert to the units specified in the user's system config.

However, there's a bit of an oddity for the forecast. 

Normally, I would handled the forecast the same way I handled the PWS observations, i.e. request Imperial, then let UoM handle the units conversion based on the user's config. But, since the narrative contains values (e.g. wind speed) as well as text (e.g. "wind gusting to 30 mph"), a mismatch of language and units can cause the narrative to be inconsistent.

In order for the narrative to contain the correct language AND value, you must call the API with the intended language AND units. Otherwise you get something like "Windböen bis 30 mph" or "Wind gusts up to 48 kph".

I've updated the comment.

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java:
```diff
@@ -0,0 +1,744 @@
+            logger.trace("Handler: Parsing forecast response: {}", response);
+            Forecast forecast = gson.fromJson(response, Forecast.class);
+            logger.debug("Handler: Successfully parsed daily forecast response object");
+            if (thing.getStatus() != ThingStatus.ONLINE) {
```
This check is not needed

mhilbush replied:
Removed

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java:
```diff
@@ -0,0 +1,744 @@
+            updateDailyForecast(forecast);
+            updateDaypartForecast(forecast.daypart);
+        } catch (JsonSyntaxException e) {
+            logger.debug("Handler: Error parsing daily forecast response object: {}", e.getMessage(), e);
```
AFAIK it will show the message of you pass the complete exception, so IMHO not needed to pass the message separately

mhilbush replied:
Message removed.

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java:
```diff
@@ -0,0 +1,744 @@
+        logger.trace("Path to icon image resource is: {}", url);
+        if (url != null) {
+            ByteArrayOutputStream out = null;
+            try {
```
Can you use a try-with-resources here?

mhilbush replied:
I've changed to try-with-resources.

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java:
```diff
@@ -0,0 +1,744 @@
+            errorMsg = String.format("InterruptedException: %s", e.getMessage());
+        }
+        markThingOffline(errorMsg);
+        logger.debug(errorMsg);
```
AFAIK this is warning for the SAT checks

mhilbush replied:
SAT isn't complaining, so I'll leave it if you're ok. Let me know if you'd prefer I change/remove it.

mhilbush replied:
After I rebased to the latest master, it started throwing errors. It's fixed now.

mhilbush commented:
@martinvw Thanks for the review. Your comments look pretty straightforward.

> Thanks for yet another contribution also here some small remarks.

I have a couple other bindings stacked up behind this one. :-) But I didn't want to submit any more PRs until things settled down with the ESH and build system migrations. The Doorbird video doorbell is closest to being ready -- just waiting for the release of a crypto library that supports the RPi. SunsetWx is pretty close, as well.

mhilbush commented:
@martinvw I believe I've addressed all your review comments. I'm going to test the new version for a while before I push the changes. I'll ping you when it's ready for review.

mhilbush commented:
@martinvw I pushed the changes to address your review comments, as well as a change to use QuantityType for rainfall rate (like I just did in the Ambient Weather binding) now that openhab/openhab-core#644 is fixed.

My approach for REFRESH was to use one channel as the trigger to invoke the backend API to get an updated forecast. That seemed to be a simpler approach than caching the results. But, I'm not sure if that approach is considered best practice for handling REFRESH. Your guidance on this would be appreciated.

martinvw commented:
> My approach for REFRESH was to use one channel as the trigger to invoke the backend API to get an updated forecast. That seemed to be a simpler approach than caching the results. But, I'm not sure if that approach is considered best practice for handling REFRESH. Your guidance on this would be appreciated.

Is that one specific channel? If you did not link/use that channel refresh is not working? Does not feel to great imho. Feel free to elaborate, but because of further obligations I (most likely) wont respond timely, I hope that one of the other @openhab/2-x-add-ons-maintainers can help you in that case 🙂

mhilbush commented:
Ah, ok. I didn't realize REFRESH happened only if the channel was linked.

I changed the implementation to cache the states, then pull from the cache on REFRESH.

openhab-5iver commented:
Thank you for this binding! Hope it gets into the distro soon!

kaikreuzer commented:
@martinvw Are your review comments all addressed? Then please add your approval.
I'd volunteer to do a quick second review.

kaikreuzer commented:
@martinvw Friendly ping!

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,135 @@
+			<description>Postal code of form postal:country (e.g. 10001:US)</description>
+		</parameter>
+
+		<parameter name="geocode" type="text" required="false">
```
Please define it just like here: https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core/src/main/resources/ESH-INF/config/i18n.xml#L55-L64
This will make the UI to offer a map selection to the user.

mhilbush replied:
Fixed.

I should note that when saving the file, Eclipse wants to replace the tabs with spaces for indentation in the CDATA section here https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core/src/main/resources/ESH-INF/config/i18n.xml#L59-L62. This causes SAT to fail the build. 🙄 I had to remove all indentation for the CDATA section to resolve the issue.

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,518 @@
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="weather">
+		<label>WeatherCompany Weather Thing</label>
```
```suggestion
		<label>WeatherCompany Weather</label>
```

mhilbush replied:
Fixed.

mhilbush replied:
Fixed

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,518 @@
+
+	<thing-type id="weather">
+		<label>WeatherCompany Weather Thing</label>
+		<description>Weather Company weather thing</description>
```
```suggestion
		<description>Weather Company weather</description>
```

mhilbush replied:
Fixed.

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,518 @@
+			<channel id="country" typeId="country" />
+			<channel id="latitude" typeId="latitude" />
+			<channel id="longitude" typeId="longitude" />
+			<channel id="elevation" typeId="elevation" />
```
you should merge these into a single channel of type "Location"

mhilbush replied:
@kaikreuzer I looked into making this change. The elevation used in `PointType` is defined to be in SI units (meters) only. As `PointType` does not use UoM, I don't understand how elevation can properly be represented in systems that are configured to use Imperial units. Maybe I'm missing something, so hopefully you can give some guidance.

kaikreuzer replied:
You are right - when introducing UoM, the `PointType` was indeed completely forgotten 😲 .
So yes, this indeed won't work if users want to use the (so outdated 😛 ) imperial units.

As the altitude is an optional value: How about merely merging latitude&longitude into a PointType channel, while keeping elevation a separate one?

mhilbush replied:
Perfect! I was going to suggest the exact same thing.

I dunno about those countries (or should I say country) that are still in the dark ages... LOL

mhilbush replied:
Completed.

mhilbush commented:
Thanks for the review @kaikreuzer!

I'll work on these. Also, I'll also add some comments/questions on your specific feedback.

mhilbush commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,2640 @@
+
+## Supported Things
+
+Only one thing type: Weather Thing
```
I did look at the WU binding. The only function of the bridge is to validate the API key and expose the API key to the thing handler. All the other work is done in the thing handler. So I thought it was a bit overkill to have a bridge just for that.

Nonetheless, if that's the recommended approach, I can rework to have a bridge. While I'm at it, I suppose it also makes sense to split the existing handler (which supports weather forecasts and personal weather station data) into a weather forecast thing and a personal weather data thing. That would make it possible to mix and match weather forecast and weather data things. Make sense?

kaikreuzer replied:
If they both share the exact same configuration (i.e. the location) and most users always require both of them, I'd say they could be kept in a single thing.
But if they are structurally different information (the one being the "own" data, the other being the predictions from other data), I also see your point. I don't think there's a wrong or right here, so please decide however you feel it is most appropriate!

mhilbush replied:
Thanks for the guidance. They do have fundamentally different data (current weather observations versus future weather forecast). And, while they both share some of the same config information, the weather data requires some additional info (e.g. the name of a weather station(s) on WeatherUnderground). My other thought was that someone might want one forecast, but might want to pull weather data from multiple weather stations. Splitting them would make it easier (and more efficient) to do that.

mhilbush replied:
This is complete.

mhilbush commented on bundles/org.openhab.binding.weathercompany/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,135 @@
+		<parameter name="language" type="text" required="true">
+			<label>Language</label>
+			<description>Language to be used by the Weather Company service</description>
+			<default>en-US</default>
```
I thought about that, but...

- the country/language pairs supported by The Weather Company are a subset of the locales that can be configured in openHAB

- during auto discovery, the binding does use the openHAB locale to initially set this language parameter

- other weather bindings (WU, OWM, DarkSky) all have language config parameters

So, I thought it was ok to have this as a parameter.

kaikreuzer replied:
ok, fine for me to keep it. But it definitely must not be marked as required :-)

mhilbush replied:
Sure. But it's ok to have a default, right?

kaikreuzer replied:
Hm. If the default is specified here in the XML, you cannot determine in your code whether the user has set it or whether it is merely the default.
Imho it is probably better to remove the default here and cover a non-set parameter in the code by first checking if the openHAB locale is one that exists in the remote API as well and if this is not the case to use something like `en-US`.

mhilbush replied:
Completed.

mhilbush commented on bundles/org.openhab.binding.weathercompany/README.md:
````diff
@@ -0,0 +1,2640 @@
+
+```
+// PWS Current Observations
+Number:Temperature          WC_PWS_Temperature
````
Fixed

mhilbush commented on bundles/org.openhab.binding.weathercompany/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,135 @@
+
+	<config-description uri="thing-type:weathercompany:weather">
+		<parameter name="apiKey" type="text" required="true">
+			<label>API Key</label>
```
Fixed

mhilbush replied:
Fixed

mhilbush commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,2640 @@
+
+## Discovery
+
+The binding will auto-discover a *Local Weather* thing if the location (i.e. latitude and longitude)
```
Completed

mhilbush commented on bundles/org.openhab.binding.weathercompany/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,135 @@
+			<description>IATA airport code (e.g. BWI)</description>
+		</parameter>
+
+		<parameter name="language" type="text" required="true">
```
Removed required flag.

mhilbush commented:
@kaikreuzer I believe I've addressed all your review feedback with the exception of [this one](https://github.com/openhab/openhab2-addons/pull/5780#discussion_r347159326), for which I would benefit from some guidance from you.

mhilbush commented:
Ok, @kaikreuzer, I think that should be everything. Ready for another review when you get a chance.

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/README.md:
````diff
@@ -0,0 +1,1040 @@
+### Thing Example
+
+```
+Bridge weathercompany:bridge:bridge [ apiKey="734982347982374" ] {
````
I hope the API key isn't a real one...

The nice thing with the account as a bridge is that you only need to declare it once:
```suggestion
Bridge weathercompany:bridge:bridge [ apiKey="734982347982374" ] {
    Thing weathercompany:weather-forecast:forecast "Local Forecast" @ "Home" [locationType="postalCode",postalCode="10001:US",language="en_US",refreshInterval=30]
    Thing weathercompany:weather-forecast:chitown "Chicago Forecast" @ "Ohare Airport" [apiKey="734982347982374",locationType="iataCode",iataCode="ORD",language="en_US",refreshInterval=30]    
    Thing weathercompany:weather-forecast:miami "Miami Weather" @ "South Beach" [locationType="postalCode",postalCode="33139:US",language="es_US",refreshInterval=30]
    Thing weathercompany:weather-observations:observations "Local Observations" @ "Home" [pwsStationId="KFLMIAMI208",refreshInterval=30]    
    Thing weathercompany:weather-observations:patagonia "Torres del Paine Weather" @ "Patagonia" [pwsStationId="IPUNTAAR4",refreshInterval=30]
}
```

mhilbush replied:
Good idea to combine together.

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,1040 @@
+Number:Length WC_Day0_PrecipitationRain "Forecasted Rainfall Amount [%.2f %unit%]" <rain> { channel="weathercompany:weather-forecast:forecast:forecastDay0#precipitationRain" }
+Number:Length WC_Day0_PrecipitationSnow "Forecasted Snowfall Amount [%.2f %unit%]" <snow> { channel="weathercompany:weather-forecast:forecast:forecastDay0#precipitationSnow" }
+
+// Day 1 - Tomorrow
```
In order to keep it readable, I would strongly suggest to really see this as an example and only show a single day in this readme. The user should be clever enough to understand that he can just copy&paste&replace the section for additional days.

mhilbush replied:
Agreed. Was trying to make it easy for users to get started, but I understand your point.

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/WeatherCompanyBindingConstants.java:
```diff
@@ -0,0 +1,140 @@
+    private static final String BINDING_ID = "weathercompany";
+
+    // Bridge
+    public static final String THING_TYPE_BRIDGE = "bridge";
```
I would have preferred a more speaking name like e.g. "account".

mhilbush replied:
Changed.

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/WeatherCompanyBridgeHandler.java:
```diff
@@ -0,0 +1,111 @@
+                    logger.debug("Bridge: HttpResponseException: API key is not valid");
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "API key is invalid");
+                } else {
+                    logger.info("Bridge: IOException trying to validate Api key: {}", e.getMessage());
```
reduce to debug

mhilbush replied:
Done

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/WeatherCompanyBridgeHandler.java:
```diff
@@ -0,0 +1,111 @@
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "API key is invalid");
+                } else {
+                    logger.info("Bridge: IOException trying to validate Api key: {}", e.getMessage());
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE, e.getMessage());
```
This is the bridge itself, so its bridge cannot be offline.
```suggestion
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, e.getMessage());
```

mhilbush replied:
Done

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,523 @@
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<bridge-type id="bridge">
+		<label>Weather Company Bridge</label>
```
```suggestion
		<label>Weather Company Account</label>
```

mhilbush replied:
Of course. Sometimes you can't see the forest for the trees... 🙄

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,523 @@
+
+	<bridge-type id="bridge">
+		<label>Weather Company Bridge</label>
+		<description>Represents a connection to the Weather Company service</description>
```
```suggestion
		<description>Represents an account at the Weather Company service</description>
```

mhilbush replied:
Done

kaikreuzer commented:
@martinvw As you were the initial reviewer, could you meanwhile please check if your comments have been addressed and provide your approval, if that's the case?

mhilbush commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,1040 @@
+**IATA Code** (required when Location Type is IATA Code):
+Three character airport code (e.g. BWI, FCO).
+
+**Langauge** (optional):
```
Done

mhilbush commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,1040 @@
+
+The following configuration parameters are available on the Bridge thing:
+
+**API Key** (required):
```
Reorganized into tables for easier reading.

mhilbush commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,1040 @@
+**Location Type** (required):
+The following location types are available:
+
+  - Postal Code
```
Added to table.

mhilbush commented:
@kaikreuzer I've addressed your latest batch of review comments.

@martinvw I know you're quite busy. I'd appreciate if you could let me know whether I adequately addressed your review comments. Thanks!!

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,349 @@
+
+| Parameter        | Parameter ID      | Required/Optional | Description |
+|------------------|-------------------|-------------------|-------------|
+| API Key          | apiKey            | Required          | Get the API key from your Weather Underground PWS page.Old Weather Underground API keys will not work with this binding. |
```
```suggestion
| API Key          | apiKey            | Required          | Get the API key from your Weather Underground PWS page. Old Weather Underground API keys will not work with this binding. |
```

mhilbush replied:
Done

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,349 @@
+
+## Supported Things
+
+The following thing types are supported:
```
This list is still lacking the technical IDs of the thing types for reference.

mhilbush replied:
Ah, ok. In all the bindings I've worked on, I don't recall ever adding the IDs of the thing types. I guess I need to update those docs, too. 😉

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,349 @@
+
+The following thing types are supported:
+
+- Account (bridge) thing for maintaining API key for accessing Weather Company API,
```
e.g. do
```suggestion
- `account`: Account for maintaining API key for accessing Weather Company API,
```

mhilbush replied:
I put then in a table.

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,349 @@
+
+## Thing Configuration
+
+### Account Thing
```
I'd leave out the "Thing" from these titles.
```suggestion
### Account
```

mhilbush replied:
Ok

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,349 @@
+|------------------|-------------------|-------------------|-------------|
+| API Key          | apiKey            | Required          | Get the API key from your Weather Underground PWS page.Old Weather Underground API keys will not work with this binding. |
+
+### Weather Forecast Thing
```
```suggestion
### Weather Forecast
```

mhilbush replied:
Ok

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/README.md:
```diff
@@ -0,0 +1,349 @@
+| Language         | language          | Optional                               | Languages supported by The Weather Company API. If the language not specified in the thing configuration, the binding will try to select the language based on the locale set in openHAB. |
+| Refresh Interval | refreshInterval   | Required                               | Frequency with which forecast will be updated. Defaults to 30 minutes. Minimum is 2 minutes. |
+
+### Weather Observations Thing
```
```suggestion
### Weather Observations
```

mhilbush replied:
Ok

mhilbush commented:
@kaikreuzer I'm done with all the changes. However, I have one more question for which I'd like your opinion...

After having created a separate thing type for the current weather observations, the channel group I had defined for the observations no longer serves any useful purpose. Do you think I should remove that channel group?

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,144 @@
+	xmlns:config-description="https://openhab.org/schemas/config-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/config-description/v1.0.0 https://openhab.org/schemas/config-description-1.0.0.xsd">
+
+	<config-description uri="thing-type:weathercompany:bridge">
```
```suggestion
	<config-description uri="thing-type:weathercompany:account">
```

mhilbush replied:
Missed that one...

kaikreuzer commented on bundles/org.openhab.binding.weathercompany/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,523 @@
+	<bridge-type id="account">
+		<label>Weather Company Account</label>
+		<description>Represents an account at the Weather Company service</description>
+		<config-description-ref uri="thing-type:weathercompany:bridge" />
```
```suggestion
		<config-description-ref uri="thing-type:weathercompany:account" />
```

mhilbush replied:
Done

martinvw commented on bundles/org.openhab.binding.weathercompany/.project:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>org.openhab.binding.bigassfan</name>
```
Please fix this name here

mhilbush replied:
Fixed

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyAbstractHandler.java:
```diff
@@ -0,0 +1,399 @@
+            dateTimeType = new DateTimeType(localDateTime);
+        } catch (DateTimeParseException e) {
+            logger.debug("Error parsing date/time string: {}", e.getMessage());
+            dateTimeType = new DateTimeType();
```
Wouldn't some `UndefType` be more appropriate here

mhilbush replied:
Yes. Agreed.

martinvw commented on bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyAbstractHandler.java:
```diff
@@ -0,0 +1,399 @@
+        } catch (DateTimeParseException e) {
+            logger.debug("Error parsing date/time string: {}", e.getMessage());
+            dateTimeType = new DateTimeType();
+        } catch (IllegalArgumentException e) {
```
In what case is this specific exception thrown? It sounds like a case which should be prevented IMHO

mhilbush replied:
IIRC, I originally was using the DateTimeType(String) constructor, which threw the IllegalArgumentException [here](https://github.com/openhab/openhab-core/blob/10acf5cf4675ba908ba69361df951bf36e9744a6/bundles/org.openhab.core.compat1x/src/main/java/org/openhab/core/library/types/DateTimeType.java#L54). However, that was not giving me the correct local time based on the openHAB timezone. I changed the implementation, but forgot to remove this catch. I think it can safely be removed.

mhilbush commented:
@martinvw I'll fix these. Just give me a couple days as I'm traveling for the next couple days.

mhilbush commented:
@martinvw I found some time tonight to work on this. Hopefully this is ready to go now.

Thanks for your help!

kaikreuzer commented:
Time to merge, I would say!

martinvw commented:
> Time to merge, I would say!

Yes! Thanks @mhilbush (and @kaikreuzer for merging)!

mhilbush commented:
Great! Thanks @kaikreuzer and @martinvw for your help!
